docs/Contributing/research/mdm/software-manage-lifecycle.md

```diff
@@ -0,0 +1,101 @@
+
+# Software Manage Lifecycle
+
+## Add Software to empty db
+
+```mermaid
+mindmap
+  Software Actions
+    Software Upload
+      Extract Version and match Title
+      Store Query for precondition
+      Store Script for Install
+      Store Script for post-install
+      Store Binary FS / S3
+      Storage Driver Configured FS
+      Storage Driver Configured S3
+    Software exists
+      API changes to show on list software
+    Delete Software
+      Remove Binary
+      Clear DB refs for scripts / query
+    Install Software to Host
+      Run Pre-condition
+      Download Binary - API?
+      Run install script
+      Run Post install script
+      Cleanup binary
+```
+
+## DB
+
+Notes:
+
+- The diagram below intends to cover everything except the orchestration part.
+- Idea:
+  - We don't need to report/filter, etc based on orchestration steps, so we don't need to track anything other than failed/success status in the db.
+  - As such, all the pre/post/rollback logic can be handled entirely by orbit
+
+```mermaid
+---
+title: DB schema
+---
+erDiagram
+  software_installers ||--|| software : "software_version"
+  software_installers ||--|| software_pre_install_conditions : "pre_install_condition_id"
+  software_installers ||--|| scripts : "install_script"
+  software_installers ||--|| scripts : "post_install_script"
+  software_installers {
+    int(10) id PK
+    bigint(20) software_version FK
+    int(10) pre_install_condition_id FK
+    int(10) install_script FK
+    int(10) post_install_script FK
+  }
+
+  software_pre_install_conditions {
+    int(10) id PK
+    text condition
+  }
+
+  host_software_pre_install_conditions ||--|| hosts : "host_id"
+  host_software_pre_install_conditions ||--|| software_pre_install_conditions : "pre_install_condition_id"
+  host_software_pre_install_conditions {
+    int(10) id PK
+    int(10) host_id FK
+    int(10) pre_install_condition_id FK
+    text output
+    varchar(20) status FK "references `host_mdm_status`"
+  }
```

Comment:
two thoughts:

1. We can remove `software_pre_install_conditions` and `host_software_pre_install_conditions` and instead:
    1. Add `pre_condition` as a column in `software_installers`
    2. Add `condition_output` and `condition_status` in `host_software_installs`

2. If we decide to keep `host_software_pre_install_conditions` we can change the foreign key `software_installers.pre_install_condition_id` to be an UUID (now or in the future) to support other conditions like scripts

Reply:
Agree with not needing the pre-install to be tables. The software installer could probably store the pre-install query directly as a column, and could use FK to `script_contents` table for the content of the install script (maybe? or this one could be inlined since it should be small) and the post-condition. 

One thing to think about, the install and post-install scripts maybe should not appear in the `scripts` table, just reference the `script_contents`? The `scripts` table will show up in the list of saved scripts, I don't think we want that.

And we should definitely add a uuid immediately to the "install software request" (I think for now this is the `host_softwrae_installs` table) so that all "activities" in the upcoming list have a uuid and can be canceled via this uuid (eventually).

Reply:
thinking a bit more about this, I like your idea of just using `script_contents` and saving the output and exit code in `host_software_installs`.

both things are scripts, but they're conceptually a different kind.

Reply:
@mna thanks so much for taking the time to answer my dumb questions.

re: `host_script_results` thanks! that makes sense. I saw that the table has a FK to `scripts.id` that's why I was thinking about just using scripts, but I think your point is that we can leave that as `NULL` and use the other FK to `script_contents.id` (which I missed)

would a good summary be, we need to define if we want:

a. some duplication (in the schema and in functionality)
b. explicitly adding filters in multiple places to hide scripts + extra logic to handle orchestration

I'm tempted to go with `a`

Reply:
>  I saw that the table has a FK to scripts.id that's why I was thinking about just using scripts, but I think your point is that we can leave that as NULL and use the other FK to script_contents.id

@roperzh yes! This is how anonymous scripts work, the `script_id` is null but the contents is not.

I think I agree with your preference of `a`. I think it would be cumbersome and error-prone to use `b`, although both options will add complexity for sure. Thinking about it a bit more, I'm wondering though if `b` would not just be to:
* Ignore those scripts when listing the upcoming queue
* Prevent recording a "past" activity when recording those scripts' results
? And of course getting a script results means doing the rest of the work for the software installer, either rollbacking or creating the installer's past activity.

Thinking out loud, but maybe b) is not _too_ complex and could be manageable, but I may be missing some other points.

Reply:
@mna that makes sense, yes especially with your suggestion to NOT use `scripts`.

The only other consideration I have is the orchestration. I was hoping we could do all the install/rollback logic in orbit, almost without server intervention something like this:

```mermaid
flowchart TB
    z("Start installation notification") --> orbit

    subgraph orbit
        direction LR
        query --> a("install script")
        a("install script") -- "maybe rollback" --> b("post install script")
    end

    orbit --> sr("Send results to Fleet")
```

If we rely on the existing script APIs, I'm imagining that we need something like:

```mermaid

flowchart TB
    z("Run query notification") --> orbit

    orbit --> a("run query, report result to server")
    
    b("server sends notification to run script")  --> c("orbit reports script results")

    d("server sends notificaton for post/install script OR rollback script") --> e("orbit reports script results")
```

And my hunch is to avoid the server orchestration, although that will give us better reporting abilities if we need them.

Reply:
Regarding `b`, are we dancing around the concept of a `script_type` ("installer", "custom", "builtin", etc.)?

Reply:
>  I was hoping we could do all the install/rollback logic in orbit, almost without server intervention
@roperzh that looks good to me, in that case I think a) makes even more sense!

@gillespi314 that was tending towards something like this, but with what Roberto is suggesting I think `a` is definitely better than `b`.

Reply:
@mna @gillespi314 thanks! I'll start with `a`, but honestly I'm not 100% convinced yet, it's nice to have things like execution_time, etc for free.

let's keep it in mind and raise concerns as we go.